src/metabase/api/card.clj

```diff
@@ -1,54 +1,150 @@
+  "Efficiently add `favorite` status for a large collection of `Cards`."
+  [cards]
+  (let [favorite-card-ids (set (sel :many :field [CardFavorite :card_id], :owner_id *current-user-id*, :card_id [in (map :id cards)]))]
+    (for [card cards]
+      (assoc card :favorite (contains? favorite-card-ids (:id card))))))
+
+(defn- cards:all
+  "Return all `Cards`."
+  []
+  (sel :many Card, :archived false, (k/order :name :ASC)))
+
+(defn- cards:mine
+  "Return all `Cards` created by current user."
+  []
+  (sel :many Card, :creator_id *current-user-id*, :archived false, (k/order :name :ASC)))
+
+(defn- cards:fav
+  "Return all `Cards` favorited by the current user."
+  []
+  (->> (hydrate (sel :many [CardFavorite :card_id], :owner_id *current-user-id*)
+                :card)
+       (map :card)
+       (filter (complement :archived))
+       (sort-by :name)))
+
+(defn- cards:database
+  "Return all `Cards` belonging to `Database` with DATABASE-ID."
+  [database-id]
+  (sel :many Card (k/order :name :ASC), :database_id database-id, :archived false))
+
+(defn- cards:table
+  "Return all `Cards` belonging to `Table` with TABLE-ID."
+  [table-id]
+  (sel :many Card (k/order :name :ASC), :table_id table-id, :archived false))
+
+(defn- cards-with-ids
+  "Return unarchived `Cards` with CARD-IDS.
+   Make sure cards are returned in the same order as CARD-IDS`; `[in card-ids]` won't preserve the order."
+  [card-ids]
+  {:pre [(every? integer? card-ids)]}
+  (let [card-id->card (sel :many :field->obj [Card :id], :id [in card-ids], :archived false)]
+    (filter identity (map card-id->card card-ids))))
+
+(defn- cards:recent
+  "Return the 10 `Cards` most recently viewed by the current user, sorted by how recently they were viewed."
+  []
+  (cards-with-ids (map :model_id (k/select ViewLog
+                                   (k/aggregate (max :timestamp) :max)
+                                   (k/group :model_id)
+                                   (k/fields :model_id)
+                                   (k/where {:model (kx/literal "card"), :user_id *current-user-id*})
+                                   (k/order :max :DESC)
+                                   (k/limit 10)))))
+
+(defn- cards:popular
+  "All `Cards`, sorted by popularity (the total number of times they are viewed in `ViewLogs`).
+   (yes, this isn't actually filtering anything, but for the sake of simplicitiy it is included amongst the filter options for the time being)."
+  []
+  (cards-with-ids (map :model_id (k/select ViewLog
+                                   (k/aggregate (count (k/raw "*")) :count)
+                                   (k/group :model_id)
+                                   (k/fields :model_id)
+                                   (k/where {:model (kx/literal "card")})
+                                   (k/order :count :DESC)))))
+
+(defn- cards:archived
+  "`Cards` that have been archived."
+  []
+  (sel :many Card, :archived true, (k/order :name :ASC)))
```

Comment:
If they require the user-id then it can be passed in as an argument.

A "persistence layer" is synonymous with the term "data access layer" and is a cornerstone of a lot of web application development architecture https://en.wikipedia.org/wiki/Data_access_layer